src/react/experimental/transition-manager.js

```diff
@@ -0,0 +1,209 @@
+
+/* eslint-disable max-depth */
+
+/* global setInterval, clearInterval */
+import assert from 'assert';
+import {
+  viewportLinearInterpolator,
+  extractViewportFrom,
+  VIEWPORT_PROPS
+} from './viewport-transition-utils';
+
+// TODO: make the frequency customizalbe , adapt to frame rate
+const VIEWPORT_TRANSITION_FREQUENCY = 0.01;
+const VIEWPORT_TRANSITION_EASING_FUNC = t => t;
+
+export const TRANSITION_EVENTS = {
+  BREAK: 1,
+  SNAP_TO_END: 2,
+  IGNORE: 3
+};
+
+const DEFAULT_PROPS = {
+  transitionDuration: 0,
+  transitionInterpolator: viewportLinearInterpolator,
+  transitionEasing: VIEWPORT_TRANSITION_EASING_FUNC,
+  transitionInterruption: TRANSITION_EVENTS.BREAK,
+  onTransitionStart: () => {},
+  onTransitionInterrupt: () => {},
+  onTransitionEnd: () => {}
+};
+
+export default class TransitionManager {
+  constructor(props, onTransitionUpdate) {
+
+    this.props = Object.assign({}, DEFAULT_PROPS, props);
+    this.onTransitionUpdate = onTransitionUpdate;
+    this.transitionContext = {
+      time: 0,
+      interval: null,
+      viewport: null,
+      startViewport: null,
+      endViewport: null
+    };
+
+    this._updateViewport = this._updateViewport.bind(this);
+    this._createTransitionInterval = this._createTransitionInterval.bind(this);
+    this._isTheUpdateDueToCurrentTransition = this._isTheUpdateDueToCurrentTransition.bind(this);
+    this._triggerTransition = this._triggerTransition.bind(this);
+    this._endTransition = this._endTransition.bind(this);
+    this._isTransitionEnabled = this._isTransitionEnabled.bind(this);
+    this._isTransitionInProgress = this._isTransitionInProgress.bind(this);
+    this.processViewportChange = this.processViewportChange.bind(this);
+    this._shouldIgnoreViewportChange = this._shouldIgnoreViewportChange.bind(this);
+  }
+
+  getTransionedViewport() {
+    return this.transitionContext.viewport;
+  }
+
+  processViewportChange(props, nextProps) {
+    // NOTE: Be cautious re-ordering statements in this function.
```

Comment:
Maybe add a few guiding comments inside the function?

Reply:
Added.